lib/tapyrus/tip0137.rb

```diff
@@ -0,0 +1,164 @@
+    # @raise [ArgumentError] raise if decoded txid is not a 64-character hexadecimal string
+    # @raise [ArgumentError] raise if decoded index is not a non-negative integer less than 2^32
+    # @raise [ArgumentError] raise if decoded color_id is an invalid Tapyrus::Color::ColorIdentifier object
+    # @raise [ArgumentError] raise if decoded value is not a non-negative integer less than 2^64
+    # @raise [ArgumentError] raise if decoded script_pubkey is an invalid Tapyrus::Script object
+    # @raise [ArgumentError] raise if decoded address is an invalid Tapyrus address
+    # @raise [ArgumentError] raise if decoded message is not a hexdecimal string
+    # @raise [ArgumentError] raise if transaction is not found with decoded txid and index in blockchain
+    # @raise [ArgumentError] raise if decoded script and value is not equal to ones in blockchain
+    def verify_message!(jws, key, client: nil)
+      JWS
+        .decode(jws, key.pubkey)
+        .tap do |decoded|
+          payload = decoded[0]
+          color_id =
+            begin
+              Tapyrus::Color::ColorIdentifier.parse_from_payload(payload['color_id']&.htb) if payload['color_id']
+            rescue => _e
+              raise ArgumentError, 'color_id is invalid'
+            end
+          script_pubkey =
+            begin
+              Tapyrus::Script.parse_from_payload(payload['script_pubkey']&.htb)
+            rescue => _e
+              raise ArgumentError, 'script_pubkey is invalid'
+            end
+          validate_payload!(
+            txid: payload['txid'],
+            index: payload['index'],
+            color_id: color_id,
+            value: payload['value'],
+            script_pubkey: script_pubkey,
+            address: payload['address'],
+            message: payload['message']
+          )
+          if client
+            validate_on_blockchain!(
+              client: client,
+              txid: payload['txid'],
+              index: payload['index'],
+              color_id: color_id,
+              value: payload['value'],
+              script_pubkey: script_pubkey
+            )
+          end
+        end
+    end
+
+    # @param jws [String] JWT Web Token
+    # @param key [Tapyrus::Key] public key
+    # @param client [Tapyrus::RPC::TapyrusCoreClient] rpc client
+    # @return [Boolean] true if JWT and decoded object is valid.
+    def verify_message(jws, key, client: nil)
+      verify_message!(jws, key, client: client)
+      true
+    rescue ArgumentError, JWT::VerificationError
+      return false
+    end
+
+    def validate_payload!(txid:, index:, value:, script_pubkey:, color_id: nil, address: nil, message: nil)
+      raise ArgumentError, 'txid is invalid' if !txid || !/^[0-9a-fA-F]{64}$/.match(txid)
+      raise ArgumentError, 'index is invalid' if !index || !/^\d+$/.match(index.to_s) || index < 0 || index >= 2**32
+      if color_id && (!color_id.is_a?(Tapyrus::Color::ColorIdentifier) || !color_id.valid?)
```

Comment:
Additional checks are required depending on the specifications below:

https://github.com/chaintope/tips/pull/5#discussion_r1405725647